src/native/libs/System.Globalization.Native/pal_icushim_static.c

```diff
@@ -98,60 +104,180 @@ static int32_t load_icu_data(void* pData)
+        }
+
+        size_t delta = old_sz - cs.dst_size;
+        dst_size += delta;
+
+        if (cs.dst_size == 0) {
+            size_t old_sz = dst_capacity;
+            size_t new_sz = (dst_capacity * 3) >> 1;
+            uint8_t *next_dst_buf = realloc(dst_buf, new_sz);
+            if (next_dst_buf == NULL) {
+                log_shim_error("apple_decompress: Failed to allocate %zu bytes when resizing decompression buffer.", new_sz);
+                goto error;
+            }
+            dst_buf = next_dst_buf;
+            dst_capacity = new_sz;
+            cs.dst_ptr = dst_buf + old_sz;
+            cs.dst_size = new_sz - old_sz;
+        }
+
+        if (cs.src_size == 0) {
+            flags = COMPRESSION_STREAM_FINALIZE;
+        }
+    }
+    compression_stream_destroy(&cs);
+    return (char *) dst_buf;
+
+error:
+    if (dst_buf != NULL) {
+        free(dst_buf);
+    }
+    if (cs_init) {
+        compression_stream_destroy(&cs);
+    }
+    return NULL;
+}
+#endif
+
 int32_t GlobalizationNative_LoadICUData(const char* path)
 {
     int32_t ret = -1;
-    char* icu_data;
+    char *icu_data = NULL;
+
+    char *file_buf = NULL;
+    long file_buf_size = 0;
+    char *uncompressed_file_buf = NULL;
 
+    int use_lzfse = 0;
+    #if defined(USE_APPLE_DECOMPRESSION)
+    {
+        size_t path_len = strlen(path);
+        use_lzfse = ends_with(path, path_len, ".lzfse", sizeof(".lzfse") - 1);
+    }
+    #else
+    #endif
     FILE *fp = fopen(path, "rb");
+    if (fp == NULL) {
+    }
+
     if (fp == NULL) {
         log_shim_error("Unable to load ICU dat file '%s'.", path);
-        return ret;
+        goto error;
     }
 
     if (fseek(fp, 0L, SEEK_END) != 0) {
-        fclose(fp);
         log_shim_error("Unable to determine size of the dat file");
-        return ret;
+        goto error;
     }
 
-    long bufsize = ftell(fp);
+    file_buf_size = ftell(fp);
 
-    if (bufsize == -1) {
-        fclose(fp);
+    if (file_buf_size == -1) {
         log_shim_error("Unable to determine size of the ICU dat file.");
-        return ret;
+        goto error;
     }
 
-    icu_data = malloc(sizeof(char) * (bufsize + 1));
+    file_buf = malloc(sizeof(char) * (file_buf_size + 1));
 
-    if (icu_data == NULL) {
-        fclose(fp);
+    if (file_buf == NULL) {
         log_shim_error("Unable to allocate enough to read the ICU dat file");
-        return ret;
+        goto error;
     }
 
     if (fseek(fp, 0L, SEEK_SET) != 0) {
-        fclose(fp);
         log_shim_error("Unable to seek ICU dat file.");
-        return ret;
+        goto error;
     }
 
-    fread(icu_data, sizeof(char), bufsize, fp);
+    fread(file_buf, sizeof(char), file_buf_size, fp);
     if (ferror( fp ) != 0 ) {
-        fclose(fp);
         log_shim_error("Unable to read ICU dat file");
-        return ret;
+        goto error;
     }
 
     fclose(fp);
+    fp = NULL;
+
+    #if defined(USE_APPLE_DECOMPRESSION)
+    if (use_lzfse) {
```

Comment:
Maybe just use strcmp and delete use_lzfse and ends_with:
```suggestion
    if (strcmp(strrchr(path, '.'), ".lzfse") == 0)
```